File: Semester_1/Labs/Lab12/external_multiphase_sort.py
import heapq
import os

def merge_files(output_file: str, temp_files: list[str], path: str) -> None:
    heap_array = []
    temp_files = [open(temp_file, 'r', encoding="utf-8")  for temp_file in temp_files]
    with open(output_file, "w") as output:
        for i in range(len(temp_files)):
            element = temp_files[i].readline().strip()
            if element:
                heapq.heappush(heap_array, (int(element), i))

        counter = 0
        while counter < len(temp_files):
            root = heapq.heappop(heap_array)
            output.write(str(root[0]) + '\n')

            element = temp_files[root[1]].readline().strip()
            if element:
                heapq.heappush(heap_array, (int(element), root[1]))
            else:
                counter += 1

        for temp_file in temp_files:
            temp_file.close()

def create_initial_runs(input_file: str, run_size: str, path: str) -> list[str]:
    temp_files = []
    with open(input_file, 'r', encoding="utf-8") as input: 
        end_of_file = False
        temp_files_counter = 0
        
        if not os.path.exists(path):
            os.makedirs(path)
        
        while True:
            data = []
            for _ in range(run_size):
                line = input.readline().strip()
                if not line:
                    end_of_file = True
                    break                    
                data.append(int(line))
            data.sort()

            with open(path + r'\f_' + str(temp_files_counter) + '.txt', 'w', encoding="utf-8") as output:
                temp_files.append(path + r'\f_' + str(temp_files_counter) + '.txt')
                output.write('\n'.join(str(i) for i in data))

            temp_files_counter += 1

            if end_of_file: break
    return temp_files


def external_multiphase_sort(path: str, run_size: int) -> None:
    #run_size - How many numbers will programm read in one run

    input_file = f"{path}/input.txt"
    output_file = f"{path}/output.txt"
    path += "\Temp_files_linear"

    temp_files = create_initial_runs(input_file, run_size, path)
    merge_files(output_file, temp_files, path)

File: Semester_1/Labs/Lab12/external_multiphase_sort.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <queue>
#include <string>
#include <algorithm>
#include <filesystem>

//! REWRITTEN FROM PYTHON USING AI, IN NEED OF TESTING AND PROBABLY WON'T WORK
//! REWRITTEN FROM PYTHON USING AI, IN NEED OF TESTING AND PROBABLY WON'T WORK
//! REWRITTEN FROM PYTHON USING AI, IN NEED OF TESTING AND PROBABLY WON'T WORK

void merge_files(const std::string& output_file, const std::vector<std::string>& temp_files, const std::string& path) {
    std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, std::greater<>> heap_array;
    std::vector<std::ifstream> temp_file_streams;
    for (const auto& temp_file : temp_files) {
        temp_file_streams.emplace_back(temp_file);
    }

    std::ofstream output(output_file);
    for (size_t i = 0; i < temp_files.size(); ++i) {
        std::string element;
        std::getline(temp_file_streams[i], element);
        if (!element.empty()) {
            heap_array.emplace(std::stoi(element), i);
        }
    }

    int counter = 0;
    while (counter < temp_files.size()) {
        auto root = heap_array.top();
        heap_array.pop();
        output << root.first << '\n';

        std::string element;
        std::getline(temp_file_streams[root.second], element);
        if (!element.empty()) {
            heap_array.emplace(std::stoi(element), root.second);
        } else {
            ++counter;
        }
    }

    for (auto& temp_file_stream : temp_file_streams) {
        temp_file_stream.close();
    }
}

std::vector<std::string> create_initial_runs(const std::string& input_file, int run_size, const std::string& path) {
    std::vector<std::string> temp_files;
    std::ifstream input(input_file);
    bool end_of_file = false;
    int temp_files_counter = 0;

    if (!std::filesystem::exists(path)) {
        std::filesystem::create_directory(path);
    }

    while (true) {
        std::vector<int> data;
        for (int i = 0; i < run_size; ++i) {
            std::string line;
            std::getline(input, line);
            if (line.empty()) {
                end_of_file = true;
                break;
            }
            data.emplace_back(std::stoi(line));
        }
        std::sort(data.begin(), data.end());

        std::string temp_file = path + "/f_" + std::to_string(temp_files_counter) + ".txt";
        temp_files.emplace_back(temp_file);
        std::ofstream output(temp_file);
        for (const auto& value : data) {
            output << value << '\n';
        }

        ++temp_files_counter;

        if (end_of_file) {
            break;
        }
    }
    return temp_files;
}

void external_multiphase_sort(const std::string& path, int run_size) {
    std::string input_file = path + "/input.txt";
    std::string output_file = path + "/output.txt";
    std::string temp_files_path = path + "/Temp_files_linear";

    std::vector<std::string> temp_files = create_initial_runs(input_file, run_size, temp_files_path);
    merge_files(output_file, temp_files, temp_files_path);
}

//! REWRITTEN FROM PYTHON USING AI, IN NEED OF TESTING AND PROBABLY WON'T WORK
//! REWRITTEN FROM PYTHON USING AI, IN NEED OF TESTING AND PROBABLY WON'T WORK
//! REWRITTEN FROM PYTHON USING AI, IN NEED OF TESTING AND PROBABLY WON'T WORK